kernel/memory/kpm.c

```diff
@@ -0,0 +1,253 @@
+	}
+
+	for (size_t n = 1; n < KPM_NORDERS; n++)
+		kpm_update_order(n, base, limit);
+}
+
+/*
+ * Set pageframes as unvailable
+ *
+ * @base: Start address of the memory region to enable
+ * @limit: Size of the region
+ *
+ */
+void kpm_disable(void *base, size_t limit) {
+	if (!ISALIGNED(base, PAGE_SIZE))
+		return;
+	if (!ISALIGNED(limit, PAGE_SIZE))
+		limit = ALIGNNEXT(limit, PAGE_SIZE);
+
+	size_t base_index = (uintptr_t)base / PAGE_SIZE;
+	for (int i = 0; i < limit / PAGE_SIZE; i++) {
+		KPM_DISABLE(base_index + i);
+		KPM_ALLOC(0, base_index + i);
+	}
+
+	for (size_t n = 1; n < KPM_NORDERS; n++)
+		kpm_update_order(n, base, limit);
+}
+
+/*
+ * Return 1 if the address is on a available page frame.
+ * Iterate while the page frame containing the address is equal to KPM_PARTENABLED,
+ * when it's done two possibility arise, the page frame is available or not.
+ *
+ * @addr: the address to check
+ *
+ */
+int kpm_isenabled(void *addr) {
+	return KPM_IS_ENABLED((uintptr_t)addr / PAGE_SIZE);
+}
+
+/*
+ * Return 1 if the address is on a allocated page frame.
+ * Iterate while the page frame containing the address is equal to KPM_PARTENABLED
+ * than if is equal to KPM_PARTALLOCATED, when it's done two possibility arise,
+ * the page frame is allocated or not.
+ *
+ * @addr: the address to check
+ *
+ */
+int kpm_isalloc(void *addr) {
+	return KPM_IS_ALLOCATED(0, (uintptr_t)addr / PAGE_SIZE);
+}
+
+static int find_best_fit_order(size_t size) {
+	size_t nframes;
+
+	nframes = ALIGNNEXT(size, PAGE_SIZE) / PAGE_SIZE;
+	for (int n = 0; n < KPM_NORDERS; n++) {
+		size_t nframes_per_block = 1 << n;
+		if (nframes_per_block >= nframes)
+			return n;
+	}
+	return KPM_NORDERS - 1;
+}
+
+/*
+ * Search for a bit set to 0 in a bitmap composed
+ * of several words.
+ *
+ * Returns the index of a bit set to 0 if found, 0 otherwise
+ */
+static int bitmap_ffu(bitmap_t *bitmap, size_t size) {
+	int ffu;
+
+	for (size_t i = 0; i < size; i++) {
+		bitmap_t *cur = bitmap + i;
+		ffu = __builtin_ffs(~*cur);
+		if (ffu)
+			return ffu + (i * 8) - 1;
+	}
+	return -1;
+}
+
+/*
+ * kpm_alloc searches for the biggest contiguous region, up to
+ * @size bytes, and fills the struct @chunk with this candidate.
+ *
+ * Returns 0 on success, -1 on error
+ *
+ * NOTE: the returned chunk may be smaller than the requested size if there
+ * is no contiguous block big enough. In this case, subsequent calls will
+ * be needed to get the remaining chunks.
+ */
+int kpm_alloc(kpm_chunk_t *chunk, size_t size) {
+	size_t best_fit_order;
+	size_t base_index;
+	size_t nframes;
+
+	best_fit_order = find_best_fit_order(size);
+	for (size_t n = best_fit_order; n > 0; n--) {
+		int i = bitmap_ffu(buddy->orders[n].bitmap, buddy->orders[n].size);
+		if (i >= 0) {
+			base_index = i * (1 << n);
+			nframes = 1 << n;
+			for (size_t j = 0; j < nframes; j++)
+				KPM_ALLOC(0, base_index + j);
+			chunk->addr = (void *)(i * PAGE_SIZE * (1 << n));
+			chunk->size = PAGE_SIZE * (1 << n);
```

Comment:
Agreed, even if it adds a bit of confusion because it's used sometimes as the number of frames we want to allocate and sometimes as the number of frames per block in the current order.
It however also makes the code more readable so... I'll replace each `1 << n` with `nframes` as you suggested, but maybe we should rename this variable. We could even have two variable assigned at the same time with two different names to makes things more clear e.g.:

```c
nframes = frames_per_block =  1 << n;
base_index = i * frames_per_block;
for (size_t j = 0; j < nframes; j++)
    KPM_ALLOC(0, base_index + j);
chunk->addr = (void *)(i * frames_per_block * PAGE_SIZE);
chunk->size = frames_per_block * PAGE_SIZE;

```

What do you think?